akka-remote/src/main/scala/akka/remote/artery/SystemMessageDelivery.scala

```diff
@@ -0,0 +1,298 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.remote.artery
+
+import java.util.ArrayDeque
+
+import scala.annotation.tailrec
+import scala.concurrent.Future
+import scala.concurrent.Promise
+import scala.concurrent.duration._
+import scala.util.Failure
+import scala.util.Success
+import scala.util.Try
+
+import akka.Done
+import akka.actor.ActorRef
+import akka.actor.Address
+import akka.remote.EndpointManager.Send
+import akka.remote.artery.Transport.InboundEnvelope
+import akka.stream.Attributes
+import akka.stream.FlowShape
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.stage.AsyncCallback
+import akka.stream.stage.GraphStage
+import akka.stream.stage.GraphStageLogic
+import akka.stream.stage.GraphStageWithMaterializedValue
+import akka.stream.stage.InHandler
+import akka.stream.stage.OutHandler
+import akka.stream.stage.TimerGraphStageLogic
+
+/**
+ * INTERNAL API
+ */
+private[akka] object SystemMessageDelivery {
+  // FIXME serialization of these messages
+  final case class SystemMessageEnvelope(message: AnyRef, seqNo: Long, ackReplyTo: ActorRef)
+  sealed trait SystemMessageReply
+  final case class Ack(seq: Long, from: Address) extends SystemMessageReply
+  final case class Nack(seq: Long, from: Address) extends SystemMessageReply
+
+  private case object ResendTick
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class SystemMessageDelivery(
+  replyJunction: SystemMessageReplyJunction.Junction,
+  resendInterval: FiniteDuration,
+  localAddress: Address,
+  remoteAddress: Address,
+  ackRecipient: ActorRef)
+  extends GraphStage[FlowShape[Send, Send]] {
+
+  import SystemMessageDelivery._
+  import SystemMessageReplyJunction._
+
+  val in: Inlet[Send] = Inlet("in")
+  val out: Outlet[Send] = Outlet("out")
+  override val shape: FlowShape[Send, Send] = FlowShape(in, out)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new TimerGraphStageLogic(shape) with InHandler with OutHandler {
+
+      var registered = false
+      var seqNo = 0L // sequence number for the first message will be 1
+      val unacknowledged = new ArrayDeque[Send]
```

Comment:
for my understanding: so this is basically the re-deliver buffer, which in current impl we bound by a given nr of messages. Will we want to keep that bound on the buffer in new impl as well?